airflow/plugins/operators/sentry_to_gcs.py

```diff
@@ -0,0 +1,164 @@
+    return issues_list
+
+
+def iterate_over_sentry_records(extract, auth_token):
+    """
+    Paginate over API responses for each targeted issue and create a combined list of dicts.
+    """
+    headers = {"Authorization": "Bearer " + auth_token}
+    issues_list = get_issues_list_from_sentry(extract, headers)
+    response_data = []
+
+    for issue_id in issues_list:
+        next_url = BASE_URL + f"issues/{issue_id}/events/"
+
+        while next_url:
+            response = requests.get(next_url, headers=headers)
+            response_data.extend(response.json())
+            if response.links["next"]["results"] == "true":
+                next_url = response.links["next"]["url"]
+            else:
+                next_url = None
+
+    return response_data
+
+
+def fetch_and_clean_from_sentry(extract, auth_token):
+    """
+    Download Sentry event records as a DataFrame.
+    """
+
+    print(f"Downloading Sentry event data for project {extract.project_slug}")
+    all_rows = iterate_over_sentry_records(extract, auth_token)
+    extract.extract_ts = pendulum.now()
+
+    raw_df = pd.DataFrame([{**make_arrays_bq_safe(row)} for row in all_rows])
+
+    cleaned_df = raw_df.rename(make_name_bq_safe, axis="columns")
+
+    extract.data = cleaned_df.to_json(orient="records", lines=True).encode()
+
+    return extract
+
+
+class SentryExtract(PartitionedGCSArtifact):
+    bucket: ClassVar[str] = CALITP_BUCKET__SENTRY_LOGS
+    table: ClassVar[str] = "events"
+    dt: pendulum.Date
+    ts: pendulum.DateTime
+    partition_names: ClassVar[List[str]] = ["dt", "ts"]
+    project_slug: str
+    data: Optional[bytes]
+    extract_ts: Optional[pendulum.DateTime]
+
+    # pydantic doesn't know dataframe type
+    # see https://stackoverflow.com/a/69200069
+    class Config:
+        arbitrary_types_allowed = True
+
+    def save_to_gcs(self, fs):
+        self.save_content(fs=fs, content=self.data, exclude={"data"})
+
+
+class SentryToGCSOperator(BaseOperator):
+
+    template_fields = ("bucket",)
+
+    def __init__(
+        self,
+        bucket,
+        project_slug,
+        auth_token=None,
+        **kwargs,
+    ):
+        """
+        An operator that downloads data from a specific issue within a Sentry instance
+        and saves it as a JSON file hive-partitioned by date and time in Google Cloud
+        Storage (GCS).
+
+        Args:
+            bucket (str): GCS bucket where the scraped Sentry issue will be saved.
+            project_slug (str): The identifier for the Sentry project being examined.
+            auth_token (str, optional): The auth token to use when downloading from Sentry.
+                This can be someone's personal auth token. If not provided, the environment
+                variable of `CALITP_SENTRY_AUTH_TOKEN` is used.
+        """
+        self.ts = pendulum.now()
+        self.dt = self.ts.date()
+        self.bucket = bucket
+        self.extract = SentryExtract(
```

Comment:
I'd call this stuff in the execute so the timestamp is accurate; technically this constructor gets called on import I think? I could be wrong